<a href="https://colab.research.google.com/github/Bhartiankita369/PW_assignment/blob/main/Advance_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import the Pandas library and load the dataset

In [6]:
import pandas as pd
df = pd.read_csv("Advanced_pandas_sales_dataset.csv")

Q1 Dataset Exploration

Perform the following exploratory steps:

Display the first 10 rows

Display the dataset structure using info()

Find the number of rows and columns

Identify the data types of each column

In [5]:
#first 10 rows
df.head(10)

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [9]:
# Dtataset structure using info
df.info()                   # 229 rows, 11 columns


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Order_ID    230 non-null    int64  
 1   Order_Date  230 non-null    object 
 2   Customer    230 non-null    object 
 3   City        230 non-null    object 
 4   Region      230 non-null    object 
 5   Product     230 non-null    object 
 6   Category    230 non-null    object 
 7   Sales       230 non-null    int64  
 8   Quantity    230 non-null    int64  
 9   Discount    230 non-null    float64
 10  Profit      230 non-null    float64
dtypes: float64(2), int64(3), object(6)
memory usage: 19.9+ KB


Q2 Sales Filtering Analysis

Filter the dataset to display:

Orders where Sales > 3500

Orders where Region = "East" and Sales > 2500

Explain what insights you observe.

# Insights
East is a high value market,
because half of the high sales orders coming from East region.

In [23]:
# Orders where Sales > 3500

high_sales = df[df["Sales"] > 3500]
print("order with sales > 3500: ")
print(high_sales)

order with sales > 3500: 
     Order_ID  Order_Date Customer       City Region   Product     Category  \
3        1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories   
4        1005  2024-10-15    Rahul  Bangalore  South   Monitor  Electronics   
7        1008  2024-03-18    Kavya    Kolkata   East     Phone  Electronics   
8        1009  2024-10-31    Kavya  Ahmedabad   East     Phone  Electronics   
10       1011  2024-06-30     Neha      Delhi  South   Printer       Office   
..        ...         ...      ...        ...    ...       ...          ...   
219      1220  2024-03-22    Meera  Bangalore   West     Phone  Electronics   
222      1223  2024-03-04    Meera  Ahmedabad  South     Phone  Electronics   
223      1224  2024-01-07      Dev  Bangalore   East    Tablet  Electronics   
225      1226  2024-06-14    Anita  Ahmedabad   East    Laptop  Electronics   
229      1230  2024-04-13    Meera    Kolkata  North   Monitor  Electronics   

     Sales  Quantity  Dis

In [68]:
# East region order with sales > 2500

east_high_sales = df[(df["Region"]=='East') & (df['Sales'] >2500)]

print("East region order with sales > 2500: ")
print(east_high_sales)

East region order with sales > 2500: 
     Order_ID  Order_Date Customer       City Region     Product     Category  \
3        1004  2024-06-06     Sara       Pune   East    Keyboard  Accessories   
7        1008  2024-03-18    Kavya    Kolkata   East       Phone  Electronics   
8        1009  2024-10-31    Kavya  Ahmedabad   East       Phone  Electronics   
13       1014  2024-03-18    Anita  Bangalore   East       Mouse  Accessories   
21       1022  2024-01-14    Nitin    Chennai   East     Monitor  Electronics   
27       1028  2024-07-21      Dev    Kolkata   East      Tablet  Electronics   
33       1034  2024-12-15    Nitin  Ahmedabad   East      Tablet  Electronics   
34       1035  2024-02-20    Nitin  Ahmedabad   East      Laptop  Electronics   
41       1042  2024-07-15    Rohan    Chennai   East      Laptop  Electronics   
43       1044  2024-09-20    Kavya  Hyderabad   East     Printer       Office   
52       1053  2024-05-24    Anita    Chennai   East       Phone  Elect

Insight : East is a high value market, because almost half of the high sales orders coming from East region.

In [69]:
print("High sales count:",len(high_sales))
print("High sales of East:",len(east_high_sales))


High sales count: 104
High sales of East: 48


Q3 Customer Purchase Analysis

Identify the top 5 customers based on total sales.

Hint: Use GroupBy with aggregation.

In [37]:
df.groupby('Customer')['Sales'].sum().sort_values(ascending=False).head()

,Sales
Customer,
Nitin,59820
Raj,58416
Rahul,52291
Dev,48500
Isha,46818


Q4 Regional Sales Performance

Calculate the total sales generated by each region.

Then identify:

Which region generated the highest revenue?


#Insights
 East is generating Highest revenue.

In [38]:
df.groupby("Region")["Sales"].sum().sort_values(ascending=False)


,Sales
Region,
East,232675
South,183108
West,170227
North,158210


Q5 Product Performance Analysis

Calculate the total quantity sold for each product.

Which product appears to be the most popular?
# Insights:
Monitor is the most popular product.

In [41]:
df.groupby("Product")["Quantity"].sum().sort_values(ascending=False)


,Quantity
Product,
Monitor,156
Phone,151
Tablet,150
Laptop,140
Mouse,122
Headphones,115
Keyboard,110
Printer,108


Q6 Multi-Level Grouping

Group the dataset by: Region and Category


Assignment Questions
Calculate the total sales for each combination.

What insights can you derive from this analysis?

# Insights:
Category Electronics is the most selling product and also East region is
highest selling of Electronics.

In [45]:
df.groupby(["Region","Category"])["Sales"].sum().sort_values(ascending=False)


,,Sales
Region,Category,
East,Electronics,160907
South,Electronics,86690
West,Electronics,84128
South,Accessories,80914
North,Electronics,70023
West,Accessories,64811
North,Accessories,62052
East,Accessories,57646
North,Office,26135


Q7 Pivot Table Creation

Create a pivot table showing:

Rows → Region

Columns → Product

Values → Total Sales

Which region sells the most laptops?

In [46]:
# Create pivot table

pivot_table = pd.pivot_table(
    df,
    values="Sales",
    index="Region",
    columns="Product",
    aggfunc="sum"
)

print(pivot_table)

Product  Headphones  Keyboard  Laptop  Monitor  Mouse  Phone  Printer  Tablet
Region                                                                       
East          23899     10845   55568    31299  22902  39667    14122   34373
North         15450     20538    4540    14922  26064  30535    26135   20026
South         24111     23480   23029    19167  33323  18218    15504   26276
West          17159     38499   10826    18178   9153  27850    21288   27274


Q8 Profitability Analysis

Calculate the average profit for each product category.

Which category is most profitable?
# Profitable category
Printer is the most profitable category

In [54]:
df.groupby("Product")["Profit"].mean().sort_values(ascending=False)


,Profit
Product,
Printer,471.938462
Phone,452.949032
Mouse,451.771111
Laptop,437.685172
Keyboard,434.287200
Tablet,397.251765
Headphones,351.856897
Monitor,293.930690


Q9 Discount Impact Study

Analyze the relationship between discount and profit.

Tasks:

Calculate the average profit for each discount level

Identify whether higher discounts lead to lower profit

-- yes, higher dicount leads to lower profit.

In [55]:
df.groupby("Discount")["Profit"].mean()

,Profit
Discount,
0.00,751.210588
0.05,531.368039
0.10,346.460513
0.15,237.191400
0.20,90.026923


Q10 End-to-End Data Wrangling Task

Perform the following workflow:

Filter the dataset where Sales > 3000

Group data by Region and Product

Calculate total sales

Reshape the result into a pivot table

Explain the business insights obtained from this analysis.

In [65]:
# Filter the dataset where Sales > 3000

print(df[df["Sales"] > 3000])

     Order_ID  Order_Date Customer       City Region   Product     Category  \
3        1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories   
4        1005  2024-10-15    Rahul  Bangalore  South   Monitor  Electronics   
5        1006  2024-07-04    Pooja  Bangalore   West   Printer       Office   
7        1008  2024-03-18    Kavya    Kolkata   East     Phone  Electronics   
8        1009  2024-10-31    Kavya  Ahmedabad   East     Phone  Electronics   
..        ...         ...      ...        ...    ...       ...          ...   
219      1220  2024-03-22    Meera  Bangalore   West     Phone  Electronics   
222      1223  2024-03-04    Meera  Ahmedabad  South     Phone  Electronics   
223      1224  2024-01-07      Dev  Bangalore   East    Tablet  Electronics   
225      1226  2024-06-14    Anita  Ahmedabad   East    Laptop  Electronics   
229      1230  2024-04-13    Meera    Kolkata  North   Monitor  Electronics   

     Sales  Quantity  Discount  Profit  
3     3933

In [63]:
# Group data by Region and Product

df.groupby(["Region","Product"])["Sales"].sum()

Region  Product   
East    Headphones    23899
        Keyboard      10845
        Laptop        55568
        Monitor       31299
        Mouse         22902
        Phone         39667
        Printer       14122
        Tablet        34373
North   Headphones    15450
        Keyboard      20538
        Laptop         4540
        Monitor       14922
        Mouse         26064
        Phone         30535
        Printer       26135
        Tablet        20026
South   Headphones    24111
        Keyboard      23480
        Laptop        23029
        Monitor       19167
        Mouse         33323
        Phone         18218
        Printer       15504
        Tablet        26276
West    Headphones    17159
        Keyboard      38499
        Laptop        10826
        Monitor       18178
        Mouse          9153
        Phone         27850
        Printer       21288
        Tablet        27274
Name: Sales, dtype: int64

In [61]:
# Reshape the result into a pivot table

pd.pivot_table(df,values="Sales",index="Region",columns="Product",aggfunc="sum")

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,23899,10845,55568,31299,22902,39667,14122,34373
North,15450,20538,4540,14922,26064,30535,26135,20026
South,24111,23480,23029,19167,33323,18218,15504,26276
West,17159,38499,10826,18178,9153,27850,21288,27274


 # Business Insights
 1.Best Performing Region Overall:
South has consistently high values across many products,
East has very strong Laptop & Phone sales.

South and East are high-revenue regions.

2. Top-Selling Products:
 "Laptop"
East: 55,568 (very high)
South: 23,029
Laptop dominates especially in East

 "Phone"
East: 39,667
West: 27,850
North: 30,535
Phones sell well across all regions

3. Weak Product Performance
 Laptop in North
Only 4,540 (very low)

 Insight:
Laptop demand is very weak in North

Mouse in West
Only 9,153


4. Balanced vs Specialized Regions
South → Balanced region (all products decent)
East → High but product-specific
North → Mixed performance
West → Strong in few categories